arg_scanner/lib/arg_scanner/state_tracker.rb

```diff
@@ -0,0 +1,145 @@
+require "set"
+require "tempfile"
+require "require_all"
+
+module ArgScanner
+  class StateTracker
+    def initialize
+      at_exit do
+        dir = ENV["ARG_SCANNER_STATE_TRACKER_DIR"]
+        dir = "." if dir.nil? || dir == ""
+        tmp_file = Tempfile.new(["classes-", ".json"], dir )
+        path = tmp_file.path
+        tmp_file.close!
+
+        begin
+          require_all Rails.root.join('lib')
+        rescue => e
+        end
+        begin
+          Rails.application.eager_load!
+        rescue => e
+        end
+
+        File.open(path,"w") { |file| print_json(file) }
+      end
+    end
+
+    private
+    def print_json(file)
+      file.puts("{")
+      print_load_path(file)
+      file.puts(",")
+      print_all_modules(file)
+      file.puts("}")
+    end
+
+    def get_constants_of_class(constants, parent, klass)
+      begin
+        constants.select {|const| parent.const_defined?(const)}.map do |const|
+          begin
+            parent.const_get(const)
+          rescue Exception => e
+            $stderr.puts(e)
+          end
+        end.select { |const| const.is_a? klass}
+      rescue => e
+        $stderr.puts(e)
+        []
+      end
+    end
+
+    def get_modules(mod)
+      get_constants_of_class(mod.constants, mod, Module)
+    end
+
+    def get_all_modules
+      queue = Queue.new
+      visited = Set.new
+      get_modules(Module).each {|mod| queue.push(mod); visited.add(mod)}
+
+      until queue.empty? do
+        mod = queue.pop
+        get_modules(mod).each do |child|
+          unless visited.include?(child)
+            queue.push(child)
+            visited.add(child)
+          end
+        end
+      end
+      visited
+    end
+
+    def array_to_json(array, stream)
```

Comment:
`http://ruby-doc.org/stdlib-2.0.0/libdoc/json/rdoc/JSON.html#method-i-dump` or its neighbours

Reply:
Unfortunately, requiring any additional modules adds new Classes/Module which was not required in the original process. Ofc, I can require `json` here, but it means that all processes under StateTracker will also have `json` classes and modules in their hierarchy. Moreover, they can add new methods into existing classes. I think dumping anything in json is not a big deal, so I personally don't want to mess original loaded modules with `json`. What do you think about it?

Reply:
1. You may collect the required data first and `require 'json'` and do the needed stuff later; since you're hooking at_exit, you won't make problems, I think :)
2. Rails uses JSON extensively so I'm pretty sure it will be available in most cases.

Generating json manually under such circumstances looks weird to me :)